functions/access.js

```diff
@@ -0,0 +1,116 @@
+const admin = require("firebase-admin");
+const functions = require("firebase-functions");
+const { HttpsError } = require("firebase-functions/lib/providers/https");
+const { passwordsMatch } = require(".");
+const { checkAuth } = require("./auth");
+const { uuidv4 } = require("uuidv4");
+
+const checkIsValidToken = async (venueId, uid, token) => {
+  const venue = await admin.collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const granted = await venue.ref.collection("accessgranted").doc(uid).get();
+  if (!granted.exists || !granted.tokens) {
+    return false;
+  }
+
+  if (Object.keys(granted.tokens).includes(token)) {
+    // Record that the token was checked
+    if (!granted.tokens[token].usedAt) {
```

Comment:
Is it possible to get an error `Cannot read .usedAt of undefined`

Reply:
I don't think so because of the check on line 18.

But let me know if you disagree!